Assets/MixedRealityToolkit.Services/InputAnimation/InputAnimation.cs

```diff
@@ -0,0 +1,710 @@
+            int iX = FindKeyframeInterval(curveX, time);
+            int iY = FindKeyframeInterval(curveY, time);
+            int iZ = FindKeyframeInterval(curveZ, time);
+            if (iX > 0 && iY > 0 && iZ > 0)
+            {
+                Vector3 v0 = new Vector3(curveX.keys[iX - 1].value, curveY.keys[iY - 1].value, curveZ.keys[iZ - 1].value);
+                Vector3 v1 = new Vector3(curveX.keys[iX].value, curveY.keys[iY].value, curveZ.keys[iZ].value);
+
+                // Merge the preceding two intervals if difference is small enough
+                if ((v1 - v0).sqrMagnitude <= sqrThreshold && (position - v1).sqrMagnitude <= sqrThreshold)
+                {
+                    curveX.RemoveKey(iX);
+                    curveY.RemoveKey(iY);
+                    curveZ.RemoveKey(iZ);
+                }
+            }
+
+            AddFloatKey(curveX, time, position.x);
+            AddFloatKey(curveY, time, position.y);
+            AddFloatKey(curveZ, time, position.z);
+        }
+
+        private static void AddRotationKeyFiltered(AnimationCurve curveX, AnimationCurve curveY, AnimationCurve curveZ, AnimationCurve curveW, float time, Quaternion rotation, float threshold)
+        {
+            float sqrThreshold = threshold * threshold;
+
+            int iX = FindKeyframeInterval(curveX, time);
+            int iY = FindKeyframeInterval(curveY, time);
+            int iZ = FindKeyframeInterval(curveZ, time);
+            int iW = FindKeyframeInterval(curveW, time);
+            if (iX > 0 && iY > 0 && iZ > 0 && iW > 0)
+            {
+                Quaternion q0 = new Quaternion(curveX.keys[iX - 1].value, curveY.keys[iY - 1].value, curveZ.keys[iZ - 1].value, curveW.keys[iW - 1].value);
+                Quaternion q1 = new Quaternion(curveX.keys[iX].value, curveY.keys[iY].value, curveZ.keys[iZ].value, curveW.keys[iW].value);
+
+                // Merge the preceding two intervals if difference is small enough
+                (q0 * Quaternion.Inverse(q1)).ToAngleAxis(out float angle0, out Vector3 axis0);
+                (rotation * Quaternion.Inverse(q0)).ToAngleAxis(out float angle1, out Vector3 axis1);
+                if (angle0 <= sqrThreshold && angle1 <= sqrThreshold)
+                {
+                    curveX.RemoveKey(iX);
+                    curveY.RemoveKey(iY);
+                    curveZ.RemoveKey(iZ);
+                    curveW.RemoveKey(iW);
+                }
+            }
+
+            // TODO make use of Bezier interpolation to allow more aggressive compression
+            // and use tangents and weights to accurately merge adjacent splines.
+            AddFloatKey(curveX, time, rotation.x);
+            AddFloatKey(curveY, time, rotation.y);
+            AddFloatKey(curveZ, time, rotation.z);
+            AddFloatKey(curveW, time, rotation.w);
+        }
+
+        /// Add a float value to an animation curve
+        private static int AddFloatKey(AnimationCurve curve, float time, float value)
```

Comment:
Nit: Might be worth adding what this returns (I think the index into the animation curve)

Same for the other things down below.

Reply:
done